data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/strategy/UnorderedConsumerRecordOffsetStrategy.java

```diff
@@ -57,64 +66,101 @@ public UnorderedConsumerRecordOffsetStrategy(final KafkaConsumer<?, ?> consumer,
   }
 
   /**
    * {@inheritDoc}
    */
   @Override
   public void successfullySentToSubscriber(final KafkaConsumerRecord<?, ?> record) {
-    // TODO evaluate if it's worth committing offsets at specified intervals per partition.
-    // commit each record
-    commit(record)
-      .onSuccess(ignored -> {
-        eventsSentCounter.increment();
-        logger.debug(
-          "committed {} {} {}",
-          keyValue("topic", record.topic()),
-          keyValue("partition", record.partition()),
-          keyValue("offset", record.offset())
-        );
-      })
-      .onFailure(cause -> logger.error(
-        "failed to commit {} {} {}",
-        keyValue("topic", record.topic()),
-        keyValue("partition", record.partition()),
-        keyValue("offset", record.offset()),
-        cause
-      ));
+    commit(record);
   }
 
   /**
    * {@inheritDoc}
    */
   @Override
   public void successfullySentToDLQ(final KafkaConsumerRecord<?, ?> record) {
-    successfullySentToSubscriber(record);
+    commit(record);
   }
 
   /**
    * {@inheritDoc}
    */
   @Override
   public void failedToSendToDLQ(final KafkaConsumerRecord<?, ?> record, final Throwable ex) {
-    // do not commit
+    mutateStateAndCheckAck(new TopicPartition(record.topic(), record.partition()), record.offset());
   }
 
   /**
    * {@inheritDoc}
    */
   @Override
   public void recordDiscarded(final KafkaConsumerRecord<?, ?> record) {
-    successfullySentToSubscriber(record);
+    mutateStateAndCheckAck(new TopicPartition(record.topic(), record.partition()), record.offset());
   }
 
-  private Future<Map<TopicPartition, OffsetAndMetadata>> commit(
-    final KafkaConsumerRecord<?, ?> record) {
-    logger.debug("committing record {}", record);
-    return consumer.commit(Map.of(
-      new TopicPartition(record.topic(), record.partition()),
-      new OffsetAndMetadata(record.offset() + 1, ""))
-    );
+  /**
+   * @return null if it shouldn't ack, otherwise the offset to ack - 1.
+   */
+  private Long mutateStateAndCheckAck(TopicPartition topicPartition, long offset) {
+    long lastAckedOffset = this.lastAckedMap.get(topicPartition); // This is always non null
+    SortedSet<Long> toAckSet = this.pendingAcksMap.computeIfAbsent(topicPartition, v -> new TreeSet<>());
+    toAckSet.add(offset);
+
+    return (toAckSet.first() == lastAckedOffset + 1 && toAckSet.last() - toAckSet.first() == toAckSet.size() - 1) ?
+      toAckSet.last() : null;
   }
+
+  private void commit(final KafkaConsumerRecord<?, ?> record) {
+    TopicPartition key = new TopicPartition(record.topic(), record.partition());
+    Long shouldAck = mutateStateAndCheckAck(key, record.offset());
+    if (shouldAck != null) {
+      // Reset the state
+      long lastAckedBeforeThisOne = this.lastAckedMap.put(key, shouldAck);
+      SortedSet<Long> messagesImGoingToAck = this.pendingAcksMap.remove(key);
```

Comment:
This should be updated on success or on failure.

Reply:
No, this is discarded in case of success, copy pasted in case of failure (if we actually need to restore the state)